game-core/src/main/java/games/strategy/triplea/delegate/battle/end/NoMoreUnits.java

```diff
@@ -0,0 +1,77 @@
+package games.strategy.triplea.delegate.battle.end;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.delegate.Matches;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Predicate;
+import lombok.Builder;
+import lombok.NonNull;
+import org.triplea.java.collections.CollectionUtils;
+
+@Builder
+public class NoMoreUnits {
+
+  public enum Winner {
+    ATTACKER,
+    DEFENDER,
+    DEFENDER_TURN_1,
+    UNDETERMINED
+  }
+
+  private @NonNull final GameData gameData;
+  private @NonNull final GamePlayer attacker;
+  private @NonNull final Collection<Unit> attackingUnits;
+  private @NonNull final Collection<Unit> defendingUnits;
+  private @NonNull final Territory battleSite;
+  private final int round;
+  private @NonNull final Runnable checkUndefendedTransports;
+  private @NonNull final Runnable checkForUnitsThatCanRollLeft;
+
+  public Winner check() {
+    final Predicate<Unit> notInfrastructure = Matches.unitIsNotInfrastructure();
+    if (CollectionUtils.getMatches(attackingUnits, notInfrastructure).isEmpty()) {
```

Comment:
You can use an `anyMatch` here to make this more performant, ie:
```
  public Winner check() {
    if(attackingUnits.stream().anyMatch(Matches.unitIsNotInfrastructure())) {
      return noAttackingUnits();
    } else if(defendingUnits.stream().anyMatch(Matches.unitIsNotInfrastructure())) {
      return noDefendingUnits();
    }
    return Winner.UNDETERMINED;
  }
```

With the update above iteration stops when you find the first unit that matches rather than iterating through all units to then check if the resulting collection is empty.

Reply:
I was just copying the original code.  I'll change it.